### PR TITLE
optimize: refactor the redis lock string to hash

### LINKED ISSUE #3008: refactor the redis data structure
## Why you need it?
 Is your feature request related to a problem? Please describe in details
目前redis模式存储数据结构都是string,存取都需要做转化，需要优化为hash

## How it could be?
A clear and concise description of what you want to happen. You can explain more about input of the feature, and output of it.

### PR DESCRIPTION
**Pull request by lightClouds917**
### Ⅰ. Describe what this PR did
redis lock data structor ，use hash to replace string 

### Ⅱ. Does this pull request fix one issue?

fixes #3008

### CODE REVIEW

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 1]
```diff
@@ -70,59 +86,78 @@ public boolean acquireLock(List<RowLock> rowLocks) {
...
-                pipeline.setnx(key, value);
+                Map<String, String> lockMap = new HashMap<>(8);
+                lockMap.put(XID,value.getXid());
+                lockMap.put(TRANSACTION_ID,value.getTransactionId().toString());
+                lockMap.put(BRANCH_ID,value.getBranchId().toString());
+                lockMap.put(RESOURCE_ID,value.getResourceId());
+                lockMap.put(TABLE_NAME,value.getTableName());
+                lockMap.put(ROW_KEY,value.getRowKey());
+                lockMap.put(PK,value.getPk());
+                pipeline.hmset(key,lockMap);
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: cannot play the role of competition lock

**Comment by a364176773** [comment]
list change hash xid branch1 branch1_info branch2 branch2_info.....
string change hash lock_table rawlock1 branch1_info rawlock2 branch2_info

del:
hgetall xid
branch1->branch1_info  json.get("rawlock")->rawlock1
hdel lock_table rawlock1

competition lock：
hsetnx lock_table rawlock1

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 2]
```diff
@@ -70,59 +87,80 @@ public boolean acquireLock(List<RowLock> rowLocks) {
...
-                pipeline.setnx(key, value);
+                Map<String, String> lockMap = new HashMap<>(8);
+                lockMap.put(XID, value.getXid());
+                lockMap.put(TRANSACTION_ID, value.getTransactionId().toString());
+                lockMap.put(BRANCH_ID, value.getBranchId().toString());
+                lockMap.put(RESOURCE_ID, value.getResourceId());
+                lockMap.put(TABLE_NAME, value.getTableName());
+                lockMap.put(ROW_KEY, value.getRowKey());
+                lockMap.put(PK, value.getPk());
+                pipeline.hmset(key, lockMap);
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:135]: cannot play the role of competition lock

Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: fixed.

**Comment by lightClouds917** [comment]
The data structure of redis lock like this now : 
 
image

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 3]
```diff
@@ -60,131 +87,179 @@
...
-                String existedValue = lockList.get(i);
-                if (existedValue == null) {
-                    String key = existedKeyList.get(i);
-                    map.put(key, JSON.toJSONString(locks.get(i)));
+            List<String> needLockKeys = new ArrayList<>();
+            needLockDOS.forEach(lockDO -> needLockKeys.add(buildLockKey(lockDO.getRowKey())));
+
+            Pipeline pipeline1 = jedis.pipelined();
+            needLockKeys.stream().forEachOrdered(
+                    needLockKey -> pipeline1.hmget(needLockKey, XID, TABLE_NAME, PK, BRANCH_ID));
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:115]: only need to xid

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 4]
```diff
@@ -60,131 +87,179 @@
...
+                pipeline.hsetnx(key, RESOURCE_ID, value.getResourceId());
+                pipeline.hsetnx(key, TABLE_NAME, value.getTableName());
+                pipeline.hsetnx(key, ROW_KEY, value.getRowKey());
+                pipeline.hsetnx(key, PK, value.getPk());
                 readyKeys.add(key);
             });
             List<Object> results = pipeline.syncAndReturnAll();
-            for (int i = 0; i < results.size(); i++) {
-                Long result = (long)results.get(i);
+            List<Integer> results2 = (List<Integer>) (List) results;
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:154]: List< Integer > results = (List<Integer>) (List) pipeline.syncAndReturnAll();

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 5]
```diff
@@ -60,131 +87,179 @@
...
-            String xidLockKey = getXidLockKey(locks.get(0).getXid());
-            Pipeline pipeline = jedis.pipelined();
-            pipeline.del(keys);
-            Arrays.stream(keys).forEach(key -> pipeline.lrem(xidLockKey, 0, key));
-            pipeline.sync();
+            Pipeline pipelined = jedis.pipelined();
+            needReleaseKeys.stream()
+                    .forEach(key -> pipelined.hmget(key, XID, TABLE_NAME, PK, BRANCH_ID));
+            List<Object> existedObjs = pipelined.syncAndReturnAll();
+            List<List<String>> existedLockInfos = (List<List<String>>) (List) existedObjs;
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:203]: the lock can only be actively unlocked by the owner without having to retrieve it again

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 3 above)
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:115]: only need to xid

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 6]
```diff
@@ -60,131 +87,179 @@
...
+
         try (Jedis jedis = JedisPooledFactory.getJedisInstance()) {
-            String xidLockKey = getXidLockKey(locks.get(0).getXid());
-            Pipeline pipeline = jedis.pipelined();
-            pipeline.del(keys);
-            Arrays.stream(keys).forEach(key -> pipeline.lrem(xidLockKey, 0, key));
-            pipeline.sync();
+            Pipeline pipelined = jedis.pipelined();
+            needReleaseKeys.stream()
+                    .forEach(key -> pipelined.hmget(key, XID, TABLE_NAME, PK, BRANCH_ID));
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:201]: only need to xid

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 7]
```diff
@@ -60,131 +87,179 @@
...
+                                existedLockInfo.get(1), existedLockInfo.get(2),
+                                existedLockInfo.get(0), existedLockInfo.get(3));
+                        return false;
+                    }
+                }
+            }
+
+            Pipeline pipelined1 = jedis.pipelined();
+            needReleaseKeys.stream().forEach(key ->
+                    pipelined1.hdel(key, XID, TRANSACTION_ID, BRANCH_ID, RESOURCE_ID,
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:221]: del needReleaseKeys. toArray (new String[0])

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 8]
```diff
@@ -60,131 +87,179 @@
...
+            if (CollectionUtils.isNotEmpty(rowKeys)) {
+                Pipeline pipelined = jedis.pipelined();
+                pipelined.hdel(xidLockKey, branchIdsArray);
+                rowKeys.stream().forEach(rowKeyStr -> {
+                    if (StringUtils.isNotEmpty(rowKeyStr)) {
+                        if (rowKeyStr.contains(SEMICOLON)) {
+                            String[] keys = rowKeyStr.split(SEMICOLON);
+                            Arrays.asList(keys).stream().forEach(rowKey -> {
+                                if (StringUtils.isNotEmpty(rowKey)) {
+                                    pipelined.hdel(rowKey, XID, TRANSACTION_ID, BRANCH_ID,
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:251]: del rowkey

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 9]
```diff
@@ -60,131 +87,179 @@
...
+                                            RESOURCE_ID, TABLE_NAME, ROW_KEY, PK);
+                                }
+                            });
+                        } else {
+                            pipelined.hdel(rowKeyStr, XID, TRANSACTION_ID, BRANCH_ID, RESOURCE_ID,
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:256]: del rowkey

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 10]
```diff
@@ -207,41 +282,23 @@ public boolean isLockable(List<RowLock> rowLocks) {
...
-                    LockDO lock = JSON.parseObject(rowlockJson, LockDO.class);
-                    if (lock != null && !Objects.equals(lock.getXid(), xid)) {
-                        return false;
-                    }
-                }
-            }
+            Pipeline pipeline = jedis.pipelined();
+            lockKeys.stream().forEach(key -> pipeline.hget(key, XID));
+            List<Object> existedRowLockXid = pipeline.syncAndReturnAll();
+            List<String> existedXids = (List<String>) (List) existedRowLockXid;
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:292]: List<String> existedXids = (List<String>) (List) pipeline.syncAndReturnAll();

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 11]
```diff
@@ -60,131 +87,178 @@
...
-                Long result = (long)results.get(i);
+            List<Integer> results2 = (List<Integer>) (List) results;
+            List<List<Integer>> partitions = Lists.partition(results2, 7);
+
+            Set<String> successSet = new HashSet<>();
+            for (int i = 0; i < partitions.size(); i++) {
                 String key = readyKeys.get(i);
-                if (result != 1) {
-                    status = result;
+                if (partitions.contains(FAILED)) {
```
[the remark this answers is not in the document]
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 12]
```diff
@@ -60,131 +87,178 @@
...
-                String[] rms = successList.toArray(new String[0]);
-                if (rms.length > 0) {
-                    jedis.del(rms);
+
+            //If someone has failed,all the lockkey which has been added need to be delete.
+            if (FAILED.equals(status)) {
+                if (successSet.size() > 0) {
+                    Pipeline pipeline2 = jedis.pipelined();
+                    successSet.forEach(locKey ->
+                            pipeline2.hdel(locKey, XID, TRANSACTION_ID, BRANCH_ID, RESOURCE_ID,
```
[the remark this answers is not in the document]
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 3 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 4 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 5 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 6 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 7 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 8 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 9 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 10 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 13]
```diff
@@ -60,131 +82,139 @@
...
-                    jedis.lpush(xidLockKey, readyKeys.toArray(new String[0]));
-                } catch (Exception e) {
-                    return false;
-                }
-                return true;
             }
+            String xidLockKey = buildXidLockKey(needLockXid);
+            StringJoiner lockKeysString = new StringJoiner(SEMICOLON);
+            needLockKeys.stream().forEach(lockKey -> lockKeysString.add(lockKey));
+            jedis.hsetnx(xidLockKey, branchId.toString(), lockKeysString.toString());
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:164]: hset enough

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 14]
```diff
@@ -207,41 +237,22 @@ public boolean isLockable(List<RowLock> rowLocks) {
...
-                    LockDO lock = JSON.parseObject(rowlockJson, LockDO.class);
-                    if (lock != null && !Objects.equals(lock.getXid(), xid)) {
-                        return false;
-                    }
-                }
-            }
+            Pipeline pipeline = jedis.pipelined();
+            lockKeys.stream().forEach(key -> pipeline.hget(key, XID));
+            List<String> existedXids = (List<String>) (List) pipeline.syncAndReturnAll();
+            return existedXids.stream().allMatch(existedXid -> xid.equals(existedXid));
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:247]: Please pay attention to this problem: list All elements are null
existedXids.size>0

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 13 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 14 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 15]
```diff
@@ -60,131 +82,139 @@
      */
     public RedisLocker() {
         logQueryLimit =
-            ConfigurationFactory.getInstance().getInt(ConfigurationKeys.STORE_REDIS_QUERY_LIMIT, DEFAULT_QUERY_LIMIT);
+                ConfigurationFactory.getInstance()
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:85]: don't need the logQueryLimit

Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

**Review by a364176773 — APPROVED** [review]
LGTM

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 16]
```diff
@@ -200,48 +217,28 @@ public boolean releaseLock(String xid, Long branchId) {
...
     }
 
-    private String getXidLockKey(String xid) {
+    private String buildXidLockKey(String xid) {
         return DEFAULT_REDIS_SEATA_LOCK_XID_PREFIX + xid;
     }
 
-    private String getLockKey(String rowKey) {
+    private String buildLockKey(String rowKey) {
         return DEFAULT_REDIS_SEATA_LOCK_PREFIX + rowKey;
```
Comment (ph3636) [inline server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:242]: Change the names of these two keys for structure error

**Review by ph3636 — APPROVED** [review]
Note that the structure is not compatible
LGTM

File: common/src/main/java/io/seata/common/Constants.java [hunk 17]
```diff
@@ -40,8 +41,15 @@
      */
     public static final String DBKEYS_SPLIT_CHAR = ",";
 
-    /** the start time of transaction */
-    public static final String START_TIME  = "start-time";
+    /**
+     * The constant ;
```
[the remark this answers is not in the document]
Comment (lightClouds917) [reply common/src/main/java/io/seata/common/Constants.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 18]
```diff
@@ -40,151 +38,170 @@
...
+        String needLockXid = rowLocks.get(0).getXid();
+        Long branchId = rowLocks.get(0).getBranchId();
+
         try (Jedis jedis = JedisPooledFactory.getJedisInstance()) {
-            List<LockDO> locks = convertToLockDO(rowLocks);
-            if (locks.size() > 1) {
-                locks =
-                    locks.stream().filter(LambdaUtils.distinctByKey(LockDO::getRowKey)).collect(Collectors.toList());
+            List<LockDO> needLockDOS = convertToLockDO(rowLocks);
+            if (needLockDOS.size() >= 1) {
```
[the remark this answers is not in the document]
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 16 above)
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: > Change the names of these two keys for structure error

finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (ph3636) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: The meaning here is the same as before, so be it

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java [hunk 19]
```diff
@@ -15,23 +15,21 @@
  */
 package io.seata.server.storage.redis.lock;
 
+import static io.seata.common.Constants.SEMICOLON;
+
+import com.google.common.collect.Lists;
```
[the remark this answers is not in the document]
Comment (ph3636) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:20]: It seemed that the java package was on the top

[the remark this answers is not in the document]
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java:18]: > It seemed that the java package was on the top

finished

File: server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/lock/RedisLocker.java]: > The meaning here is the same as before, so be it

finished

**Review by l81893521 — APPROVED** [review]
LGTM for @ph3636

**Pull request merged**